pkg/volume/csi_migration.go

```diff
@@ -0,0 +1,119 @@
+limitations under the License.
+*/
+
+package volume
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/api/core/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/component-base/featuregate"
+	csilibplugins "k8s.io/csi-translation-lib/plugins"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+// InTreeToCSITranslator contains methods required to check migratable status
+// and perform translations from InTree PV's to CSI
+type InTreeToCSITranslator interface {
+	GetInTreePluginNameFromSpec(pv *v1.PersistentVolume, vol *v1.Volume) (string, error)
+	TranslateInTreePVToCSI(pv *v1.PersistentVolume) (*v1.PersistentVolume, error)
+	TranslateInTreeInlineVolumeToCSI(vol *v1.Volume) (*v1.PersistentVolume, error)
+}
+
+// CSIMigratedPluginManager keeps track of migrated state of in-tree plugins
+type CSIMigratedPluginManager struct{
+	translator InTreeToCSITranslator
+}
+
+// NewCSIMigratedPluginManager returns a new CSIMigratedPluginManager instance
+func NewCSIMigratedPluginManager(t InTreeToCSITranslator) CSIMigratedPluginManager {
+	return CSIMigratedPluginManager{
+		translator: t,
+	}
+}
+
+func isFeatureFlagEnabled(key featuregate.Feature) bool {
+	return utilfeature.DefaultFeatureGate.Enabled(key)
+}
+
+// IsCSIMigrationEnabledForCluster returns a boolean value indicating whether
+// CSI migration feature flags (either CSIMigration or CSIMigrationInTreeOff)
+// is enabled on the cluster as a whole
+func (pm CSIMigratedPluginManager) IsCSIMigrationEnabledForCluster() bool {
+	return isFeatureFlagEnabled(features.CSIMigration) || isFeatureFlagEnabled(features.CSIMigrationInTreeOff)
+}
+
+// IsCSIMigrationEnabledForPluginByName returns a boolean value indicating whether
+// the CSI migration has been enabled for a particular storage plugin cluster-wide
+func (pm CSIMigratedPluginManager) IsCSIMigrationEnabledForPluginByName(pluginName string) bool {
+	// In-tree storage to CSI driver migration feature should be enabled,
+	// along with the plugin-specific one
+	if !pm.IsCSIMigrationEnabledForCluster() {
+		return false
+	}
+
+	switch pluginName {
+	case csilibplugins.AWSEBSInTreePluginName:
+		return isFeatureFlagEnabled(features.CSIMigrationAWS)
+	case csilibplugins.GCEPDInTreePluginName:
+		return isFeatureFlagEnabled(features.CSIMigrationGCE)
+	case csilibplugins.AzureFileInTreePluginName:
+		return isFeatureFlagEnabled(features.CSIMigrationAzureFile)
+	case csilibplugins.AzureDiskInTreePluginName:
+		return isFeatureFlagEnabled(features.CSIMigrationAzureDisk)
+	case csilibplugins.CinderInTreePluginName:
+		return isFeatureFlagEnabled(features.CSIMigrationOpenStack)
+	default:
+		return false
+	}
+}
+
+func (pm CSIMigratedPluginManager) IsPluginMigratableToCSIBySpec(spec *Spec) (bool, error) {
+	if spec == nil {
+		return false, fmt.Errorf("could not find if plugin is migratable because volume spec is nil")
+	}
+
+	pluginName, _ := pm.translator.GetInTreePluginNameFromSpec(spec.PersistentVolume, spec.Volume)
```

Comment:
so I think this call is fine, we can just reframe it as instead of `translator` it's actually an `MigrationInTreePluginNameGetter` interface